internal/uvm/scsi.go

```diff
@@ -44,39 +44,35 @@ const (
 const scsiCurrentSerialVersionID = 2
 
 var (
-	ErrNoAvailableLocation      = fmt.Errorf("no available location")
-	ErrNotAttached              = fmt.Errorf("not attached")
-	ErrAlreadyAttached          = fmt.Errorf("already attached")
-	ErrNoSCSIControllers        = fmt.Errorf("no SCSI controllers configured for this utility VM")
-	ErrTooManyAttachments       = fmt.Errorf("too many SCSI attachments")
-	ErrSCSILayerWCOWUnsupported = fmt.Errorf("SCSI attached layers are not supported for WCOW")
+	ErrNoAvailableLocation    = fmt.Errorf("no available location")
+	ErrNotAttached            = fmt.Errorf("not attached")
+	ErrAlreadyAttached        = fmt.Errorf("already attached")
+	ErrNoSCSIControllers      = fmt.Errorf("no SCSI controllers configured for this utility VM")
+	ErrMoreMountsThanExpected = fmt.Errorf("SCSI attachment has more mounts than expected")
 )
 
-// Release frees the resources of the corresponding Scsi Mount
 func (sm *SCSIMount) Release(ctx context.Context) error {
-	if err := sm.vm.RemoveSCSI(ctx, sm.HostPath); err != nil {
+	if err := sm.vm.RemoveSCSIMount(ctx, sm.HostPath, sm.UVMPath); err != nil {
 		return fmt.Errorf("failed to remove SCSI device: %s", err)
 	}
 	return nil
 }
 
-// SCSIMount struct representing a SCSI mount point and the UVM
+// SCSIAttachment struct representing a SCSI mount point and the UVM
 // it belongs to.
-type SCSIMount struct {
-	// Utility VM the scsi mount belongs to
-	vm *UtilityVM
+type SCSIAttachment struct {
 	// path is the host path to the vhd that is mounted.
 	HostPath string
-	// path for the uvm
-	UVMPath string
+	mounts   map[string]*SCSIMount
 	// scsi controller
 	Controller int
 	// scsi logical unit number
 	LUN int32
 	// While most VHDs attached to SCSI are scratch spaces, in the case of LCOW
 	// when the size is over the size possible to attach to PMEM, we use SCSI for
-	// read-only layers. As RO layers are shared, we perform ref-counting.
-	isLayer  bool
+	// read-only layers.
+	isLayer bool
+	// ref count the attachment so we know when to remove it
 	refCount uint32
```

Comment:
it might be cleaner/simpler to use `len(sa.mounts)` as the ref-count, since that way we detect if an attachment is being used by the number of mounts it still has, and not need to also increment (decrement) the refcount ever time we add (remove) a mount
(That does require calling `delete(sa.mounts)` when we remove a mount though)